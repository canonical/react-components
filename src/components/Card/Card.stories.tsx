import React from "react";
import type { Meta } from "@storybook/react";

import Card from "./";
import Col from "../Col";
import Row from "../Row";

const meta: Meta<typeof Card> = {
  component: Card,
  tags: ["autodocs"],
};

export default meta;

export const Default = {
  args: {
    title: "We'd love to have you join us as a partner.",
    children:
      "If you are an independent software vendor or bundle author, it's easy to apply. You can find out more below.",
  },
};

export const Header = {
  args: {
    title: "Raspberry Pi2 and Pi3",
    thumbnail: "https://assets.ubuntu.com/v1/dca2e4c4-raspberry-logo.png",
    children:
      "For fun, for education and for profit, the RPi makes device development personal and entertaining. With support for both the Pi2 and the new Pi3, Ubuntu Core supports the worldâ€™s most beloved board.",
  },
};

export const Highlighted = {
  args: {
    ...Default.args,
    highlighted: true,
  },
};

export const Overlay = {
  args: {
    title: "Web browsing",
    overlay: true,
    children: `Renowned for speed and security, Ubuntu and Firefox make browsing
the web a pleasure again. Ubuntu also includes Chrome, Opera and
other browsers that can be installed from the Ubuntu Software
Centre.`,
  },
  render: (args) => (
    <section
      className="p-strip--image is-light"
      style={{
        backgroundImage:
          "url('https://assets.ubuntu.com/v1/0a98afcd-screenshot_desktop.jpg')",
      }}
    >
      <Row>
        <Col size={6} emptyLarge={7}>
          <Card {...args} />
        </Col>
      </Row>
    </section>
  ),
};
