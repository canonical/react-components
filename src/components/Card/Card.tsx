import classNames from "classnames";
import React from "react";
import type { HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread } from "types";
import { useId } from "hooks";

export type Props = PropsWithSpread<
  {
    /**
     * The content of the card.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to pass to the wrapping div element.
     */
    className?: ClassName;
    /**
     * Whether the card should have highlighted styling.
     */
    highlighted?: boolean;
    /**
     * Whether the card should have overlay styling.
     */
    overlay?: boolean;
    /**
     * The path to a thumbnail image.
     */
    thumbnail?: string;
    /**
     * The title of the card.
     */
    title?: ReactNode;
  },
  HTMLProps<HTMLDivElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Card](https://docs.vanillaframework.io/patterns/card/).
 *
 * There are four card styles available to use in Vanilla: default, header, highlighted and overlay. Our card component will expand to fill the full width of its parent container.
 */

const Card = ({
  children,
  className,
  highlighted,
  overlay,
  thumbnail,
  title,
  ...props
}: Props): JSX.Element => {
  const titleId = useId();
  return (
    <div
      aria-labelledby={title ? titleId : undefined}
      className={classNames(className, {
        "p-card": !highlighted && !overlay,
        "p-card--highlighted": highlighted,
        "p-card--overlay": overlay,
      })}
      role="group"
      {...props}
    >
      {thumbnail && (
        <>
          <img className="p-card__thumbnail" src={thumbnail} alt="" />
          <hr className="u-sv1" />
        </>
      )}
      {title && (
        <h3 className="p-card__title" id={titleId}>
          {title}
        </h3>
      )}
      <div className="p-card__content">{children}</div>
    </div>
  );
};

export default Card;
