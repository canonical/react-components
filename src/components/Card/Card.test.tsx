import { render, screen } from "@testing-library/react";
import React from "react";

import Card from "./Card";

describe("Card ", () => {
  it("renders", () => {
    render(<Card title="This is the title">Test content</Card>);
    expect(screen.get).toMatchSnapshot();
  });

  it("can display a header", () => {
    render(<Card thumbnail="test.png"></Card>);
    expect(wrapper.find(".p-card__thumbnail").prop("src")).toEqual("test.png");
  });

  it("can be highlighted", () => {
    render(<Card highlighted={true}></Card>);
    expect(wrapper.prop("className").includes("p-card--highlighted")).toBe(
      true
    );
  });

  it("can be an overlay", () => {
    render(<Card overlay={true}></Card>);
    expect(wrapper.prop("className").includes("p-card--overlay")).toBe(true);
  });

  it("can add additional classes", () => {
    render(<Card className="extra-class"></Card>);
    const className = wrapper.prop("className");
    expect(className.includes("p-card")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });
});
