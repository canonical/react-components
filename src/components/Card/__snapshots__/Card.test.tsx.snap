// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card  renders 1`] = `
<div
  aria-labelledby="mock-nanoid-1"
  class="p-card"
  role="group"
>
  <h3
    class="p-card__title"
    id="mock-nanoid-1"
  >
    This is the title
  </h3>
  <div
    class="p-card__content"
  >
    Test content
  </div>
</div>
`;
